# **Financial Institutions Consumer Complaints**

---
Authors:

# 2. Data Understanding
> In this stage, the study shall seek to understand the attributes of the data.

The Consumer Complaint Database is a compilation of complaints submitted to businesses asking them to address consumer financial goods and services. Complaints are publicized after the business responds and verifies a business relationship with the customer or after 15 days, whichever happens first. The Consumer Complaint Database does not include complaints that have been forwarded to other regulators, such as those regarding depository institutions with less than $10 billion in assets.

## 2.1 Data Description

## 2.2 Load the Data
> In this stage we loaded the data.

In [4]:
# Loading dependencies
import pandas as pd
import numpy as np

In [5]:
# Loading the data
def load_data(filename:str):
    """ A function that loads data"""
    output = pd.read_csv(filename)
    return output
data = load_data('data/complaints.csv')

# previewing the data
data.head()

c:\Users\Teofilo_Ligawa\anaconda3\envs\learn-env\lib\site-packages\IPython\core\interactiveshell.py:3337: DtypeWarning: Columns (9,16) have mixed types.Specify dtype option on import or set low_memory=False.
  if (await self.run_code(code, result,  async_=asy)):


Date received                                            Product  \
0    2022-12-04                        Checking or savings account   
1    2022-12-09  Credit reporting, credit repair services, or o...   
2    2022-12-09  Credit reporting, credit repair services, or o...   
3    2022-12-09  Credit reporting, credit repair services, or o...   
4    2022-12-04  Credit reporting, credit repair services, or o...   

                        Sub-product  \
0  Other banking product or service   
1                  Credit reporting   
2                  Credit reporting   
3                  Credit reporting   
4                  Credit reporting   

                                               Issue  \
0  Problem with a lender or other company chargin...   
1  Problem with a credit reporting company's inve...   
2                        Improper use of your report   
3  Problem with a credit reporting company's inve...   
4                        Improper use of your report   

                                           Sub-issue  \
0                     Transaction was not authorized   
1  Their investigation did not fix an error on yo...   
2      Reporting company used your report improperly   
3         Problem with personal statement of dispute   
4      Reporting company used your report improperly   

  Consumer complaint narrative  \
0                          NaN   
1                          NaN   
2                          NaN   
3                          NaN   
4                          NaN   

                             Company public response  \
0  Company has responded to the consumer and the ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4  Company has responded to the consumer and the ...   

                                 Company State ZIP code Tags  \
0  BANK OF AMERICA, NATIONAL ASSOCIATION    RI     2832  NaN   
1                          EQUIFAX, INC.    TX    79928  NaN   
2                          EQUIFAX, INC.    GA    30103  NaN   
3                          EQUIFAX, INC.    PA    19104  NaN   
4    Experian Information Solutions Inc.    FL    33167  NaN   

  Consumer consent provided? Submitted via Date sent to company  \
0                      Other           Web           2022-12-04   
1                        NaN           Web           2022-12-09   
2                        NaN           Web           2022-12-09   
3                        NaN           Web           2022-12-09   
4                      Other           Web           2022-12-04   

  Company response to consumer Timely response? Consumer disputed?  \
0      Closed with explanation              Yes                NaN   
1                  In progress              Yes                NaN   
2                  In progress              Yes                NaN   
3                  In progress              Yes                NaN   
4      Closed with explanation              Yes                NaN   

   Complaint ID  
0       6277013  
1       6300442  
2       6300444  
3       6300449  
4       6274454

In [6]:
# Previewing the tail
data.tail()

Date received           Product                  Sub-product  \
3147565    2017-02-09   Debt collection                I do not know   
3147566    2015-04-29          Mortgage  Conventional fixed mortgage   
3147567    2017-03-31  Credit reporting                          NaN   
3147568    2017-01-16  Credit reporting                          NaN   
3147569    2018-03-07          Mortgage       Other type of mortgage   

                                            Issue  \
3147565     Cont'd attempts collect debt not owed   
3147566  Loan modification,collection,foreclosure   
3147567  Credit reporting company's investigation   
3147568    Incorrect information on credit report   
3147569            Trouble during payment process   

                                        Sub-issue  \
3147565         Debt resulted from identity theft   
3147566                                       NaN   
3147567  No notice of investigation status/result   
3147568                            Account status   
3147569                                       NaN   

                              Consumer complaint narrative  \
3147565  I have disputed my debts several times with no...   
3147566  My father died in XX/XX/XXXX. Left me his only...   
3147567  cfbp i would Like to file a complaint on Exper...   
3147568  My husband and I are in the middle of an FHA S...   
3147569                                                NaN   

                                   Company public response  \
3147565                                                NaN   
3147566                                                NaN   
3147567  Company has responded to the consumer and the ...   
3147568  Company has responded to the consumer and the ...   
3147569  Company has responded to the consumer and the ...   

                                        Company State ZIP code           Tags  \
3147565      Bonneville Billing and Collections    UT    84054  Servicemember   
3147566                          CITIBANK, N.A.    OK    74066            NaN   
3147567     Experian Information Solutions Inc.    MN    55379            NaN   
3147568  TRANSUNION INTERMEDIATE HOLDINGS, INC.    GA    30215            NaN   
3147569                   WELLS FARGO & COMPANY    CA    91304            NaN   

        Consumer consent provided? Submitted via Date sent to company  \
3147565           Consent provided           Web           2017-02-09   
3147566           Consent provided           Web           2015-04-29   
3147567           Consent provided           Web           2017-03-31   
3147568           Consent provided           Web           2017-01-16   
3147569                        NaN      Referral           2018-03-08   

            Company response to consumer Timely response? Consumer disputed?  \
3147565          Closed with explanation              Yes                 No   
3147566          Closed with explanation              Yes                 No   
3147567  Closed with non-monetary relief              Yes                Yes   
3147568          Closed with explanation              Yes                 No   
3147569          Closed with explanation              Yes                NaN   

         Complaint ID  
3147565       2334969  
3147566       1352738  
3147567       2412926  
3147568       2292586  
3147569       2837068

## 2.3 Preliminary Data Inspection

In [7]:
## Checking the data summary

# Checking the dimensions of the data
print(f"This data has {data.shape[0]} rows and {data.shape[1]} columns")
print()
# summary
data.info()

This data has 3147570 rows and 18 columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3147570 entries, 0 to 3147569
Data columns (total 18 columns):
 #   Column                        Dtype 
---  ------                        ----- 
 0   Date received                 object
 1   Product                       object
 2   Sub-product                   object
 3   Issue                         object
 4   Sub-issue                     object
 5   Consumer complaint narrative  object
 6   Company public response       object
 7   Company                       object
 8   State                         object
 9   ZIP code                      object
 10  Tags                          object
 11  Consumer consent provided?    object
 12  Submitted via                 object
 13  Date sent to company          object
 14  Company response to consumer  object
 15  Timely response?              object
 16  Consumer disputed?            object
 17  Complaint ID                  int64 
dtyp

### 2.3.1 Observations:
* The columns `Date received` and `Date sent to company` have been registered as object which is basic pandas for string yet they should be date_time objects.
* Every other data type looks okay.
* From the summary above, we could note that the data has 3147570 rows and 18 columns.
* There is only one numeric column - `Complaint ID`. This column in unnecessarily integer because it is a unique identifier of the consumer.

* The other columns are objects, which is pandas for string.

In [8]:
# Creating a class to check missing values
class understanding(object):
    """ This is a class that checks for missing values"""
    def __init__(self):
        pass
    
    def miss_no(df):
        """ A function that counts the missing values per column"""
        return df.isna().sum()
   
    def percent_missing(df):
        """ A function that calculates the percentage of the column that is missing"""
        return df.isna().sum() / len(df)
    def check_dup(df):
        """ A function that checks for duplicates in the data"""
        return df.duplicated().sum()
    
    def counts(df, col):
        """ A function that finds the value counts of a column"""
        return df[col].value_counts()
    
    def num_unique(df, col):
        """A function that finds the number of unique elements in  a column"""
        return df[col].nunique()
    
    def get_unique(df, col):
        """ A function that shows the unique values in a column"""
        return df[col].unique()

In [9]:
# missing values
understand = understanding

# The number of missing values per column
understand.miss_no(data)

Date received                         0
Product                               0
Sub-product                      235293
Issue                                 0
Sub-issue                        685381
Consumer complaint narrative    2015226
Company public response         1774590
Company                               0
State                             40190
ZIP code                          40631
Tags                            2792441
Consumer consent provided?       826205
Submitted via                         0
Date sent to company                  0
Company response to consumer          4
Timely response?                      0
Consumer disputed?              2379130
Complaint ID                          0
dtype: int64

In [10]:
# Percentage of missing values per column
understand.percent_missing(data)

Date received                   0.000000
Product                         0.000000
Sub-product                     0.074754
Issue                           0.000000
Sub-issue                       0.217749
Consumer complaint narrative    0.640248
Company public response         0.563797
Company                         0.000000
State                           0.012769
ZIP code                        0.012909
Tags                            0.887174
Consumer consent provided?      0.262490
Submitted via                   0.000000
Date sent to company            0.000000
Company response to consumer    0.000001
Timely response?                0.000000
Consumer disputed?              0.755862
Complaint ID                    0.000000
dtype: float64

There are a number of missing values in the data. Most notably, the columns with a high percentage of missing values are:
* `Consumer complaint narrative` - 64 %.
* `Company public response` - 56 %.
* `Tags` - 89 %.
* `Consumer disputed?` - 76%.

In [11]:
# The number of duplicated rows in the data
understand.check_dup(data)

0

There are no duplicates in the data.

### 2.3.2 Checking Selected Columns

#### 2.3.2.1 `Consumer  disputed?`

In [12]:
# Checking the value counts of the column 'Consumer disputed?'
understand.counts(data, 'Consumer disputed?')

No     620062
Yes    148378
Name: Consumer disputed?, dtype: int64

This is a binary variable

The classes seem to be imbalanced with a significant difference in the value counts of the classes.

#### 2.3.2.2 `Timely response?`

In [13]:
# Checking the value counts of the column 'Timely response?'
understand.counts(data, 'Timely response?')

Yes    3097148
No       50422
Name: Timely response?, dtype: int64

This is a binary variable.

The classes seem to be imbalanced with a significant difference in the value counts of the classes.

#### 2.3.2.3 `Company response to consumer`

In [14]:
# Checking the value counts of the column 'Company response to consumer'
understand.counts(data, 'Company response to consumer')

Closed with explanation            2429039
Closed with non-monetary relief     474084
Closed with monetary relief         123817
In progress                          70517
Closed without relief                17868
Closed                               17611
Untimely response                     9326
Closed with relief                    5304
Name: Company response to consumer, dtype: int64

This column is a multi class column.

The classes seem to be imbalanced with a significant difference in the value counts of the classes.

#### 2.3.2.4 `Submitted via`

In [15]:
# Checking the value counts of the 'Submitted via'
understand.counts(data, 'Submitted via')

Web             2649097
Referral         237715
Phone            147848
Postal mail       86339
Fax               25660
Web Referral        487
Email               424
Name: Submitted via, dtype: int64

This column is a multi class column.

The classes seem to be imbalanced with a significant difference in the value counts of the classes.

#### 2.3.2.5 `Product`

In [16]:
# Product
understand.counts(data, 'Product')

Credit reporting, credit repair services, or other personal consumer reports    1469228
Debt collection                                                                  455016
Mortgage                                                                         366006
Credit card or prepaid card                                                      167334
Checking or savings account                                                      143480
Credit reporting                                                                 140429
Credit card                                                                       89190
Bank account or service                                                           86206
Student loan                                                                      71858
Money transfer, virtual currency, or money service                                48365
Vehicle loan or lease                                                             37604
Consumer Loan                   

#### 2.3.2.6 `Sub-product`

In [17]:
# Sub-product
understand.counts(data, 'Sub-product')

Credit reporting                              1452711
Checking account                               172737
General-purpose credit card or charge card     127050
I do not know                                   99542
Other mortgage                                  86635
                                               ...   
Credit repair                                     102
Traveler’s/Cashier’s checks                        88
Transit card                                       37
Student prepaid card                               26
Electronic Benefit Transfer / EBT card             12
Name: Sub-product, Length: 76, dtype: int64

#### 2.3.2.7 `Issue`

In [18]:
# Issue value counts
understand.counts(data, 'Issue')

Incorrect information on your report                                                772860
Problem with a credit reporting company's investigation into an existing problem    393166
Improper use of your report                                                         278791
Attempts to collect debt not owed                                                   154126
Loan modification,collection,foreclosure                                            112309
                                                                                     ...  
Property was sold                                                                        9
Lender damaged or destroyed vehicle                                                      8
Lender sold the property                                                                 7
Property was damaged or destroyed property                                               6
Lender damaged or destroyed property                                                     3

#### 2.3.2.8 `Sub-issue`

In [19]:
# Sub issue value counts
understand.counts(data, 'Sub-issue')

Information belongs to someone else                                                 496049
Their investigation did not fix an error on your report                             212930
Reporting company used your report improperly                                       176426
Credit inquiries on your report that you don't recognize                             99760
Account status incorrect                                                             90651
                                                                                     ...  
Was signed up for overdraft on card, but don't want to be                                6
Insurance terms                                                                          6
Problem with a credit reporting company's investigation into an existing problem         3
Improper use of your report                                                              1
Credit monitoring or identity theft protection services                                  1

There are a number of columns that look like they provide similiar information, such as:
* Product and Sub-product
* Issue and Sub-issue

# 3. Data Preparation
> In this phase, the data is prepared based on findings.

## 3.1 Validity
> Placing Correct formats

In [20]:
# Changing the data type of the date columns that had been noted
for col in ['Date received', 'Date sent to company']:
    data[col] = pd.to_datetime(data[col])
    
# previewing the changes
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3147570 entries, 0 to 3147569
Data columns (total 18 columns):
 #   Column                        Dtype         
---  ------                        -----         
 0   Date received                 datetime64[ns]
 1   Product                       object        
 2   Sub-product                   object        
 3   Issue                         object        
 4   Sub-issue                     object        
 5   Consumer complaint narrative  object        
 6   Company public response       object        
 7   Company                       object        
 8   State                         object        
 9   ZIP code                      object        
 10  Tags                          object        
 11  Consumer consent provided?    object        
 12  Submitted via                 object        
 13  Date sent to company          datetime64[ns]
 14  Company response to consumer  object        
 15  Timely response?              ob

## 3.2 Consistency
The data did not have any duplicates.

## 3.3 Uniformity
> Renaming the columns

In [21]:
# creating a variables of previous names
previous_names = data.columns.tolist()

# creating a variable of new names
def zipper(df):
    """ A function that zips the previous names to the new names"""
    # Listing the names
    previous_names = data.columns.tolist()
    
    # Removing whitespace
    new_names = [item.lower().replace(" ", "_") for item in data.columns.tolist()]
    
    # Removing hyphens
    new_names = [item.replace("-", "_") for item in new_names]
    
    # Removing question marks
    new_names = [item.strip("?") for item in new_names]
    
    # Creating the dictionary
    final_dict = dict(zip(previous_names, new_names))
    
    return final_dict
# Creating a dictionary of previous and new_names
names_dict = zipper(data)

# Previewing the dictionary of names
names_dict

{'Date received': 'date_received',
 'Product': 'product',
 'Sub-product': 'sub_product',
 'Issue': 'issue',
 'Sub-issue': 'sub_issue',
 'Consumer complaint narrative': 'consumer_complaint_narrative',
 'Company public response': 'company_public_response',
 'Company': 'company',
 'State': 'state',
 'ZIP code': 'zip_code',
 'Tags': 'tags',
 'Consumer consent provided?': 'consumer_consent_provided',
 'Submitted via': 'submitted_via',
 'Date sent to company': 'date_sent_to_company',
 'Company response to consumer': 'company_response_to_consumer',
 'Timely response?': 'timely_response',
 'Consumer disputed?': 'consumer_disputed',
 'Complaint ID': 'complaint_id'}

In [22]:
# renaming the columns to a uniform state
data.rename(columns=names_dict, inplace=True)

# Previewing the columns
data.columns.tolist()

['date_received',
 'product',
 'sub_product',
 'issue',
 'sub_issue',
 'consumer_complaint_narrative',
 'company_public_response',
 'company',
 'state',
 'zip_code',
 'tags',
 'consumer_consent_provided',
 'submitted_via',
 'date_sent_to_company',
 'company_response_to_consumer',
 'timely_response',
 'consumer_disputed',
 'complaint_id']

## 3.4 Completeness
> Removing missing values

In [23]:
# Percentage of missing values
understand.percent_missing(data)

date_received                   0.000000
product                         0.000000
sub_product                     0.074754
issue                           0.000000
sub_issue                       0.217749
consumer_complaint_narrative    0.640248
company_public_response         0.563797
company                         0.000000
state                           0.012769
zip_code                        0.012909
tags                            0.887174
consumer_consent_provided       0.262490
submitted_via                   0.000000
date_sent_to_company            0.000000
company_response_to_consumer    0.000001
timely_response                 0.000000
consumer_disputed               0.755862
complaint_id                    0.000000
dtype: float64

Columns with a high percentage (majority of the column) of missing values will be dropped, except if the column is a prospective target variable.

The study shall use the non-missing value of the column `consumer_disputed` as the new data frame.

In [24]:
# Selecting non-null values of the column 'consumer_disputed'
df = data.loc[data['consumer_disputed'].notnull()]

# prevewing the percentage of missing values in the new data frame
understand.percent_missing(df)

date_received                   0.000000
product                         0.000000
sub_product                     0.306025
issue                           0.000000
sub_issue                       0.592633
consumer_complaint_narrative    0.786506
company_public_response         0.745316
company                         0.000000
state                           0.007353
zip_code                        0.007349
tags                            0.858806
consumer_consent_provided       0.612354
submitted_via                   0.000000
date_sent_to_company            0.000000
company_response_to_consumer    0.000000
timely_response                 0.000000
consumer_disputed               0.000000
complaint_id                    0.000000
dtype: float64

It is worth noting that the `consumer_complaint_narrative` has several missing values yet it looks useful.
We shall fill the missing values with an empty string. The reason behind this is because some customers may have filed a complaint without providing a narrative for the complaint. It does not necessarily mean that was no complaint by the customer.

The columns that still contain a significant number of missing values are:
* `consumer_consent_provided`.
* `tags`.
* `company_public_response`.
* `sub_issue`.
* `sub_product`.

Missing values of columns will be dropped on the grounds that some columns have too many missing value and the others (`sub_product` and `sub_issue`) provide similar information to `product` and `issue` respectively.

The rest of the missing values will be dropped row-wise.

`complaint_id` shall be dropped because it doesn't add more value past unique identification.


In [25]:
# dropping columns that have been mentioned above
df = df.drop(['consumer_consent_provided', 'tags',
                'company_public_response', 'sub_issue',
                'sub_product', 'complaint_id'], axis=1)

# Previewing the data
df

date_received           product  \
76176      2015-05-07          Mortgage   
81724      2012-07-02       Credit card   
84314      2017-04-20      Student loan   
84679      2016-06-01          Mortgage   
84695      2014-08-11          Mortgage   
...               ...               ...   
3147564    2015-07-14       Credit card   
3147565    2017-02-09   Debt collection   
3147566    2015-04-29          Mortgage   
3147567    2017-03-31  Credit reporting   
3147568    2017-01-16  Credit reporting   

                                            issue  \
76176    Loan servicing, payments, escrow account   
81724                              Payoff process   
84314          Dealing with my lender or servicer   
84679    Loan servicing, payments, escrow account   
84695                Settlement process and costs   
...                                           ...   
3147564                        Delinquent account   
3147565     Cont'd attempts collect debt not owed   
3147566  Loan modification,collection,foreclosure   
3147567  Credit reporting company's investigation   
3147568    Incorrect information on credit report   

                              consumer_complaint_narrative  \
76176    1. Mortgage sold by XXXX XXXX XXXX to the XXXX...   
81724                                                  NaN   
84314    they keep raising my interest rate and i 'm co...   
84679    My Mortgage Company, Ditech, failed to apply p...   
84695                                                  NaN   
...                                                    ...   
3147564  Automated calls from " XXXX with Capital One '...   
3147565  I have disputed my debts several times with no...   
3147566  My father died in XX/XX/XXXX. Left me his only...   
3147567  cfbp i would Like to file a complaint on Exper...   
3147568  My husband and I are in the middle of an FHA S...   

                                        company state zip_code submitted_via  \
76176                       NATIONSTAR MORTGAGE    CA    90028           Web   
81724                             DISCOVER BANK    NC    27513           Web   
84314                   Navient Solutions, LLC.    CT     6770           Web   
84679                      Ditech Financial LLC    SC    29414           Web   
84695               Ocwen Financial Corporation    WI    54220           Web   
...                                         ...   ...      ...           ...   
3147564       CAPITAL ONE FINANCIAL CORPORATION    TX    77584           Web   
3147565      Bonneville Billing and Collections    UT    84054           Web   
3147566                          CITIBANK, N.A.    OK    74066           Web   
3147567     Experian Information Solutions Inc.    MN    55379           Web   
3147568  TRANSUNION INTERMEDIATE HOLDINGS, INC.    GA    30215           Web   

        date_sent_to_company     company_response_to_consumer timely_response  \
76176             2015-05-07          Closed with explanation             Yes   
81724             2012-07-03      Closed with monetary relief             Yes   
84314             2017-04-20          Closed with explanation             Yes   
84679             2016-06-01          Closed with explanation             Yes   
84695             2014-08-11          Closed with explanation             Yes   
...                      ...                              ...             ...   
3147564           2015-07-14          Closed with explanation             Yes   
3147565           2017-02-09          Closed with explanation             Yes   
3147566           2015-04-29          Closed with explanation             Yes   
3147567           2017-03-31  Closed with non-monetary relief             Yes   
3147568           2017-01-16          Closed with explanation             Yes   

        consumer_disputed  
76176                 Yes  
81724                  No  
84314                  No  
84679                 Yes  
84695                  No  
...    

In [26]:
# Checking percentage of missing values
understand.percent_missing(df)

date_received                   0.000000
product                         0.000000
issue                           0.000000
consumer_complaint_narrative    0.786506
company                         0.000000
state                           0.007353
zip_code                        0.007349
submitted_via                   0.000000
date_sent_to_company            0.000000
company_response_to_consumer    0.000000
timely_response                 0.000000
consumer_disputed               0.000000
dtype: float64

In [27]:
# Replacing missing value of 'consumer_complaint_narrative' with a string
df['consumer_complaint_narrative'].fillna('', inplace=True)

# Checking percentage of missing values again - confirming changes
understand.percent_missing(df)

date_received                   0.000000
product                         0.000000
issue                           0.000000
consumer_complaint_narrative    0.000000
company                         0.000000
state                           0.007353
zip_code                        0.007349
submitted_via                   0.000000
date_sent_to_company            0.000000
company_response_to_consumer    0.000000
timely_response                 0.000000
consumer_disputed               0.000000
dtype: float64

Missing values of `state` and `zip_code` can be dropped row-wise

In [28]:
# Dropping the missing values
df.dropna(inplace=True)

# Checking percentage of missing values again - confirming changes
understand.percent_missing(df)

date_received                   0.0
product                         0.0
issue                           0.0
consumer_complaint_narrative    0.0
company                         0.0
state                           0.0
zip_code                        0.0
submitted_via                   0.0
date_sent_to_company            0.0
company_response_to_consumer    0.0
timely_response                 0.0
consumer_disputed               0.0
dtype: float64

In [29]:
# Checking the dimensions
print(f"Before cleaning: this data has {data.shape[0]} rows and {data.shape[1]} columns")
print(f"After cleaning: this data has {df.shape[0]} rows and {df.shape[1]} columns")

Before cleaning: this data has 3147570 rows and 18 columns
After cleaning: this data has 762774 rows and 12 columns


## 3.5 Feature Engineering
> In this phase we shall generate new columns from the existing ones.

In [30]:
# Subtracting to get number of days
df['days_taken'] = df['date_sent_to_company'] - df['date_received']

# Previewing the changes
df

date_received           product  \
76176      2015-05-07          Mortgage   
81724      2012-07-02       Credit card   
84314      2017-04-20      Student loan   
84679      2016-06-01          Mortgage   
84695      2014-08-11          Mortgage   
...               ...               ...   
3147564    2015-07-14       Credit card   
3147565    2017-02-09   Debt collection   
3147566    2015-04-29          Mortgage   
3147567    2017-03-31  Credit reporting   
3147568    2017-01-16  Credit reporting   

                                            issue  \
76176    Loan servicing, payments, escrow account   
81724                              Payoff process   
84314          Dealing with my lender or servicer   
84679    Loan servicing, payments, escrow account   
84695                Settlement process and costs   
...                                           ...   
3147564                        Delinquent account   
3147565     Cont'd attempts collect debt not owed   
3147566  Loan modification,collection,foreclosure   
3147567  Credit reporting company's investigation   
3147568    Incorrect information on credit report   

                              consumer_complaint_narrative  \
76176    1. Mortgage sold by XXXX XXXX XXXX to the XXXX...   
81724                                                        
84314    they keep raising my interest rate and i 'm co...   
84679    My Mortgage Company, Ditech, failed to apply p...   
84695                                                        
...                                                    ...   
3147564  Automated calls from " XXXX with Capital One '...   
3147565  I have disputed my debts several times with no...   
3147566  My father died in XX/XX/XXXX. Left me his only...   
3147567  cfbp i would Like to file a complaint on Exper...   
3147568  My husband and I are in the middle of an FHA S...   

                                        company state zip_code submitted_via  \
76176                       NATIONSTAR MORTGAGE    CA    90028           Web   
81724                             DISCOVER BANK    NC    27513           Web   
84314                   Navient Solutions, LLC.    CT     6770           Web   
84679                      Ditech Financial LLC    SC    29414           Web   
84695               Ocwen Financial Corporation    WI    54220           Web   
...                                         ...   ...      ...           ...   
3147564       CAPITAL ONE FINANCIAL CORPORATION    TX    77584           Web   
3147565      Bonneville Billing and Collections    UT    84054           Web   
3147566                          CITIBANK, N.A.    OK    74066           Web   
3147567     Experian Information Solutions Inc.    MN    55379           Web   
3147568  TRANSUNION INTERMEDIATE HOLDINGS, INC.    GA    30215           Web   

        date_sent_to_company     company_response_to_consumer timely_response  \
76176             2015-05-07          Closed with explanation             Yes   
81724             2012-07-03      Closed with monetary relief             Yes   
84314             2017-04-20          Closed with explanation             Yes   
84679             2016-06-01          Closed with explanation             Yes   
84695             2014-08-11          Closed with explanation             Yes   
...                      ...                              ...             ...   
3147564           2015-07-14          Closed with explanation             Yes   
3147565           2017-02-09          Closed with explanation             Yes   
3147566           2015-04-29          Closed with explanation             Yes   
3147567           2017-03-31  Closed with non-monetary relief             Yes   
3147568           2017-01-16          Closed with explanation             Yes   

        consumer_disputed days_taken  
76176                 Yes     0 days  
81724                  No     1 days  
84314                  No     0 days  
84679              

In [31]:
# Getting value counts of 'days_taken'
understand.counts(df, 'days_taken')

0 days      367823
1 days       84933
2 days       58175
3 days       45710
4 days       43334
             ...  
412 days         1
601 days         1
310 days         1
686 days         1
397 days         1
Name: days_taken, Length: 488, dtype: int64

In [32]:
# Converting to integer
df['days_taken'] = df['days_taken'].astype('str').apply(lambda x: x.strip(' days')).astype('int')

# Previewing the data
df

date_received           product  \
76176      2015-05-07          Mortgage   
81724      2012-07-02       Credit card   
84314      2017-04-20      Student loan   
84679      2016-06-01          Mortgage   
84695      2014-08-11          Mortgage   
...               ...               ...   
3147564    2015-07-14       Credit card   
3147565    2017-02-09   Debt collection   
3147566    2015-04-29          Mortgage   
3147567    2017-03-31  Credit reporting   
3147568    2017-01-16  Credit reporting   

                                            issue  \
76176    Loan servicing, payments, escrow account   
81724                              Payoff process   
84314          Dealing with my lender or servicer   
84679    Loan servicing, payments, escrow account   
84695                Settlement process and costs   
...                                           ...   
3147564                        Delinquent account   
3147565     Cont'd attempts collect debt not owed   
3147566  Loan modification,collection,foreclosure   
3147567  Credit reporting company's investigation   
3147568    Incorrect information on credit report   

                              consumer_complaint_narrative  \
76176    1. Mortgage sold by XXXX XXXX XXXX to the XXXX...   
81724                                                        
84314    they keep raising my interest rate and i 'm co...   
84679    My Mortgage Company, Ditech, failed to apply p...   
84695                                                        
...                                                    ...   
3147564  Automated calls from " XXXX with Capital One '...   
3147565  I have disputed my debts several times with no...   
3147566  My father died in XX/XX/XXXX. Left me his only...   
3147567  cfbp i would Like to file a complaint on Exper...   
3147568  My husband and I are in the middle of an FHA S...   

                                        company state zip_code submitted_via  \
76176                       NATIONSTAR MORTGAGE    CA    90028           Web   
81724                             DISCOVER BANK    NC    27513           Web   
84314                   Navient Solutions, LLC.    CT     6770           Web   
84679                      Ditech Financial LLC    SC    29414           Web   
84695               Ocwen Financial Corporation    WI    54220           Web   
...                                         ...   ...      ...           ...   
3147564       CAPITAL ONE FINANCIAL CORPORATION    TX    77584           Web   
3147565      Bonneville Billing and Collections    UT    84054           Web   
3147566                          CITIBANK, N.A.    OK    74066           Web   
3147567     Experian Information Solutions Inc.    MN    55379           Web   
3147568  TRANSUNION INTERMEDIATE HOLDINGS, INC.    GA    30215           Web   

        date_sent_to_company     company_response_to_consumer timely_response  \
76176             2015-05-07          Closed with explanation             Yes   
81724             2012-07-03      Closed with monetary relief             Yes   
84314             2017-04-20          Closed with explanation             Yes   
84679             2016-06-01          Closed with explanation             Yes   
84695             2014-08-11          Closed with explanation             Yes   
...                      ...                              ...             ...   
3147564           2015-07-14          Closed with explanation             Yes   
3147565           2017-02-09          Closed with explanation             Yes   
3147566           2015-04-29          Closed with explanation             Yes   
3147567           2017-03-31  Closed with non-monetary relief             Yes   
3147568           2017-01-16          Closed with explanation             Yes   

        consumer_disputed  days_taken  
76176                 Yes           0  
81724                  No           1  
84314                  No           0  
84679          